Plot Editing: Simple & Advanced mode + Undo & Redo

vandenman opened the pull request:
Fixes most of https://github.com/jasp-stats/INTERNAL-jasp/issues/1191
Requires: https://github.com/jasp-stats/jaspGraphs/pull/10


Some things that are not fixed:
- All plots that can't be edited should have a disabled menu entry with a tooltip explaining why it can't be edited.
- Duplicate analysis doesn't copy the state and causes errors.
- Tooltips don't work when a QML thingy is disabled (see the TextField example).

Aside from that, discrete axes needs some work, but that might require a new QML component and/ or different approach.

vandenman committed:
started on simple and advanced mode

vandenman committed:
restructured all logic to C++

vandenman committed:
can now save image from inside ploteditor

vandenman committed:
actually emit the signal

vandenman committed:
started on undo mechanism

vandenman committed:
remove png files from qrc

vandenman committed:
redo undo function but values in shown in qml are not yet updated

vandenman committed:
undo and redo now functional, icons could use some love though

vandenman committed:
simplify if possible when enabling advanced mode + some fixes to advanced mode

vandenman committed:
remove old log statements and commented blocks

JorisGoosen commented on Desktop/components/JASP/Widgets/PlotEditor/PlotEditor.qml:
```diff
@@ -81,61 +81,68 @@ Popup
+							id:				xAxis
+							title:			qsTr("x-axis")
+							axisModel:		plotEditorModel.xAxis
+							visible:		axisDropDown.value === "x-axis"
```
It would be better to add an Enum for the axis and select it based on that.

JorisGoosen replied:
To make the selected axis better available here you could add a property to `PlotEditorModel` called `selectedAxis` and have that be a reference to the `PlotEditorAxisModel*` that is current.

But maybe it is better if I implement that.

vandenman replied:
I changed it, let me know if you have any further comments!

JorisGoosen commented on Desktop/results/ploteditoraxismodel.cpp:
```diff
@@ -399,20 +466,22 @@ std::string		AxisModel::LimitsTypeToString(LimitsType type) const
 }
 
-AxisModel::LimitsType	AxisModel::LimitsTypeFromString(std::string type) const
+AxisModel::LimitsType	AxisModel::LimitsTypeFromString(const std::string& type) const
```
Ah yeah, oops

JorisGoosen commented on Desktop/results/ploteditormodel.cpp:
```diff
@@ -134,6 +143,86 @@ void PlotEditorModel::somethingChanged()
+	emit undoEnabledChanged(_undoEnabled);
+}
+
+void PlotEditorModel::setRedoEnabled(bool redoEnabled)
```
This whole function isn't necessary and I'll show you later how it can be done more efficiently.

JorisGoosen replied:
(And avoiding double administration, because redo-enabled == redostacksize > 0 )

JorisGoosen reviewed (changes requested):
Alright, overall this looks good.
Two things:
- [x] You forgot to add undo/redo buttons to the darkTheme... 
- [x] If you have the "Manual" breaks selected in simple mode and you switch to advanced it selects something else.

Now, I can have a look at that last issue and also clean up the axisselection.
I would like you to add those icons though.

JorisGoosen commented:
I'll change the axis-enum and the redo enabled later if you like.

JorisGoosen commented:
(Of course if you do not like I can help you figure how to do those changes)

vandenman committed:
implemented some of the feedback of Joris

JorisGoosen committed:
Simplify undo redo enabled changed properties
- use stack directly
- no need for two separate signals, they are always emitted at the same time anyway

JorisGoosen commented on Desktop/mainwindow.cpp:
```diff
@@ -168,7 +168,8 @@ MainWindow::MainWindow(QApplication * application) : QObject(application), _appl
 	qmlRegisterType<ResultsJsInterface>							("JASP",		1, 0, "ResultsJsInterface"				);
 
-	qmlRegisterUncreatableType<PlotEditor::AxisModel>			("JASP.PlotEditor",		1, 0, "AxisModel",		"Can't make it");
+	qmlRegisterUncreatableType<PlotEditor::AxisModel>			("JASP.PlotEditorAxis",		1, 0, "AxisModel",			"Can't make it");
```
Isn't it more logical to put both of these QML items inside the same group called "JASP.PlotEditor" ?

That way you can just do `import JASP.PlotEditor 1.0` and then use `AxisModel` and `PlotEditorModel` directly.
The way you've done it now is rather confusing.

vandenman replied:
that somehow didn't work before, but I'll try again :)

vandenman replied:
and now it does work ...

vandenman committed:
Implement feedback + polish icons

JorisGoosen reviewed (approved):
Ok, this is starting to look good.
I will be merging it so other people can play around with it.

But there are some improvements we can (and should) still make.
